xds/internal/balancer/ringhash/e2e/ringhash_balancer_test.go

```diff
@@ -949,18 +907,654 @@ func (s) TestRingHash_EndpointWeights(t *testing.T) {
+		t.Fatalf("Failed to update xDS resources: %v", err)
+	}
+
+	conn, err := grpc.NewClient("xds:///test.server", grpc.WithResolvers(xdsResolver), grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		t.Fatalf("Failed to create client: %s", err)
+	}
+	defer conn.Close()
+	client := testgrpc.NewTestServiceClient(conn)
+
+	if got, want := conn.GetState(), connectivity.Idle; got != want {
+		t.Errorf("conn.GetState(): got %v, want %v", got, want)
+	}
+
+	checkRPCSendOK(ctx, t, client, 1)
+
+	if got, want := conn.GetState(), connectivity.Ready; got != want {
+		t.Errorf("conn.GetState(): got %v, want %v", got, want)
+	}
+}
+
+// Test that the channel will transition to READY once it starts
+// connecting even if there are no RPCs being sent to the picker.
+func (s) TestRingHash_ContinuesConnectingWithoutPicks(t *testing.T) {
+	backend := stubserver.StartTestService(t, &stubserver.StubServer{
+		// We expect the server EmptyCall to not be call here because the
+		// aggregated channel state is never READY when the call is pending.
+		EmptyCallF: func(ctx context.Context, _ *testpb.Empty) (*testpb.Empty, error) {
+			t.Errorf("EmptyCall() should not have been called")
+			return &testpb.Empty{}, nil
+		},
+	})
+	defer backend.Stop()
+
+	nonExistantServerAddr := makeNonExistentBackends(t, 1)[0]
+
+	const clusterName = "cluster"
+	endpoints := endpointResource(t, clusterName, []string{backend.Address, nonExistantServerAddr})
+	cluster := e2e.ClusterResourceWithOptions(e2e.ClusterOptions{
+		ClusterName: clusterName,
+		ServiceName: clusterName,
+		Policy:      e2e.LoadBalancingPolicyRingHash,
+	})
+	route := headerHashRoute("new_route", virtualHostName, clusterName, "address_hash")
+	listener := e2e.DefaultClientListener(virtualHostName, route.Name)
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+
+	xdsServer, nodeID, xdsResolver := setupManagementServerAndResolver(t)
+	if err := xdsServer.Update(ctx, xdsUpdateOpts(nodeID, endpoints, cluster, route, listener)); err != nil {
+		t.Fatalf("Failed to update xDS resources: %v", err)
+	}
+
+	dialer := testutils.NewBlockingDialer()
+	dopts := []grpc.DialOption{
+		grpc.WithResolvers(xdsResolver),
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+		grpc.WithContextDialer(dialer.DialContext),
+	}
+	conn, err := grpc.NewClient("xds:///test.server", dopts...)
+	if err != nil {
+		t.Fatalf("Failed to create client: %s", err)
+	}
+	defer conn.Close()
+	client := testgrpc.NewTestServiceClient(conn)
+
+	rpcCtx, rpcCancel := context.WithCancel(ctx)
+	go func() {
+		rpcCtx = metadata.NewOutgoingContext(rpcCtx, metadata.Pairs("address_hash", nonExistantServerAddr+"_0"))
+		_, err := client.EmptyCall(rpcCtx, &testpb.Empty{})
+		if status.Code(err) != codes.Canceled {
+			t.Errorf("Expected RPC to be canceled, got error: %v", err)
+		}
+	}()
+
+	// Wait for the connection attempt to the real backend.
+	hold := dialer.Hold(backend.Address)
```

Comment:
Should we create this hold before we spawn the goroutine that makes the RPC? If not, we could have a case where the connection attempt to this address is made before we get here, right?

Reply:
Good catch, thank you -- this should be flaky otherwise. Fixed by move the hold creation above spawning the goroutine for sending an RPC.